New spirv_new test case OpSpecConstant

gwawiork opened the pull request:
OpSpecConstant OpSpecConstantTrue OpSpecConstantfalse
Fixes #718

kpet commented on test_conformance/spirv_new/main.cpp:
```diff
@@ -81,7 +81,12 @@ void spirvTestsRegistry::addTestClass(baseTestClass *test, const char *testName)
     testDef.func = test->getFunction();
     testDef.name = testName;
-    testDef.min_version = Version(2, 1);
+    if (strstr(testName, "op_spec_constant_")) {
```
This is a nasty hack :). We will need to change the test registration logic to be able to pass an optional min version. I'd like to reuse some of that registration logic a bit more widely to reduce the amount of boilerplate so I'd like to keep it reasonably clean. I suggest doing this as a separate pre-requisite patch.

gwawiork replied:
Ok thanks. Yes it is nasty but works :D.

kpet commented:
This needs rebasing / merging master into to fix the CI.

kpet commented on test_conformance/spirv_new/spirv_asm/op_spec_constant_uint_simple.spvasm32:
```diff
@@ -0,0 +1,74 @@
+               OpGroupDecorate %37 %numbers_buffer_addr %spec_constant_value_addr %id
+       %uint = OpTypeInt 32 0
+      %uchar = OpTypeInt 8 0
+     %uint_0 = OpConstant %uint 0
```
I'm guessing you intended this to be a spec constant. It would be good to understand why the test passed (it's not really fit for purpose if it lets that through).

gwawiork replied:
Sorry this is my mistake. The test was failing. I fixed 64 bit version but forgot about 32. Thanks for catching this.

StuartDBrady replied:
@kpet Wondering what your views on this are now. It's a lot better than it was, although there's still some room for improvement. Would it be okay to address this separately, once we've removing the binaries, do you think?

kpet replied:
@StuartDBrady Assuming you intended to reply to that comment https://github.com/KhronosGroup/OpenCL-CTS/pull/719#pullrequestreview-412082959 and not this thread, I think we should indeed address that after removing the binaries. Getting forward progress on this PR is definitely more important IMHO.

kpet commented on test_conformance/spirv_new/test_op_spec_constant.cpp:
Why pass the version since you're always passing the default value? Why was the change to the test registration code needed?

gwawiork replied:
This is my typo. This should be Version(2,2). In comparison to other test cases which requires Version(2,1). This is why I need modify registration code (this solves your comment https://github.com/KhronosGroup/OpenCL-CTS/pull/719#discussion_r402922370)

gwawiork commented:
> I've had more of a look at the tests. I'm concerned that these tests will be hard to maintain due to the sheer volume of duplication across the SPIR-V assembly files. This is not something new but we should start thinking of a solution to this (most of these are just variations where only one type changes). I'm not saying we absolutely need to solve this for this change but we should think of how we could mitigate that, for example:
> 
> * Could we shorten the code?
> * Do we really need all the `OpName`?

Thank you for review of this code and spending your time. Sure something here could be improved. However from my point of view this is additional work and should be done as a additional issue. Regarding OpName I think I can remove them.

kpet commented on test_conformance/spirv_new/test_op_spec_constant.cpp:
Not saying this needs to change but: why is this a vector? This will run 1024 times the same test. I'm not sure there's a lot of value in that. We could have much simpler SPIR-V if we changed these tests to operate on a single value.

gwawiork replied:
From validation experience more is better then less :). If you insist on change I could prepare.

gwawiork commented:
> You could have kept _some_ of the OpNames for readability, I was asking a question, not requesting that you remove all of them really :). Sorry if that wasn't clear.

If you do not mind I would leave this as is now :/

StuartDBrady commented on test_conformance/spirv_new/test_op_spec_constant.cpp:
This can fail with CL_INVALID_OPERATION, because there are already kernel objects attached to the program. Please use a new program, _without_ a call to `clBuildProgram()` between the call to `clCreateProgramWithIL()` and the call to `clSetProgramSpecializationConstant()`.

gwawiork replied:
Could you clarify? My understanding is that changing value by clSetProgramSpecializationConstant requires program rebuilding using clBuildProgram. You can read about this [here](https://www.khronos.org/registry/OpenCL/specs/3.0-unified/html/OpenCL_API.html#_setting_spir_v_specialization_constants). Noticed that clSetProgramSpecializationConstant is run second time.

gwawiork replied:
In my opinion there is no need to create program object once again. However I create once again kernel object and clSetKernelArg calls are executed

StuartDBrady replied:
> Could you clarify? My understanding is that changing value by clSetProgramSpecializationConstant requires program rebuilding using clBuildProgram. You can read about this [here](https://www.khronos.org/registry/OpenCL/specs/3.0-unified/html/OpenCL_API.html#_setting_spir_v_specialization_constants). Noticed that clSetProgramSpecializationConstant is run second time.

`clSetProgramSpecializationConstant()` must be called _before_ `clBuildProgram()`. `clBuildProgram()` is called by `get_program_with_il()`.

> In my opinion there is no need to create program object once again.

The [specification states](https://www.khronos.org/registry/OpenCL/specs/3.0-unified/html/OpenCL_API.html#clBuildProgram) that `clBuildProgram()` returns:
> * `CL_​INVALID_​OPERATION` if there are kernel objects attached to _program_.

The test must not call `clBuildProgram()` for a program after calling `clCreateKernel()` for that same program.

> However I create once again kernel object and clSetKernelArg calls are executed

It is not clear to me what you are saying, here. Are you simply describing what the code in the PR currently does? If so, I've read the code and understand what is written: there's no real need to turn it into English, for me, at least.

gwawiork replied:
Looks that second usage the objects I've made is forbidden.  Then I will create new objects. Thanks

StuartDBrady commented on test_conformance/spirv_new/test_op_spec_constant.cpp:
This `spec_buffer` argument does not seem to be used for anything, in the kernels that I have looked at. Could you explain its purpose, please, or remove it if it has none?

gwawiork replied:
Good catch. It looks really not needed.

StuartDBrady commented on test_conformance/spirv_new/test_op_spec_constant.cpp:
It is of no real value to have separate signed and unsigned versions of the tests for each size of integer type, because signedness is not visible within SPIR-V, nor is it visible to the `clSetProgramSpecializationConstant()` API function. Please could you remove _either_ the signed tests _or_ the unsigned tests?

StuartDBrady replied:
Note: "_each size_ of integer type". Could you please do this for the other SPIR-V Integer types, i.e. 8-bit (uchar), 16-bit (ushort) and 32-bit (ulong), too? Thanks.

StuartDBrady commented on test_conformance/spirv_new/spirv_asm/op_spec_constant_char_simple.spvasm32:
```diff
@@ -0,0 +1,69 @@
+         %26 = OpLoad %uint %id Aligned 4
+   %arrayidx = OpInBoundsPtrAccessChain %_ptr_CrossWorkgroup_uchar %25 %26
+         %28 = OpLoad %uchar %arrayidx Aligned 1
+       %conv = OpSConvert %uint %28
+         %30 = OpLoad %_ptr_CrossWorkgroup_uchar %spec_constant_value_addr Aligned 4
+         %31 = OpLoad %uchar %30 Aligned 1
+      %conv1 = OpSConvert %uint %31
+        %add = OpIAdd %uint %conv %conv1
+      %conv2 = OpUConvert %uchar %add
```
This could be simplified by removing the `OpSConvert` and `OpUConvert` instructions.

StuartDBrady commented on test_conformance/spirv_new/spirv_asm/op_spec_constant_short_simple.spvasm32:
```diff
@@ -0,0 +1,71 @@
+         %27 = OpLoad %uint %id Aligned 4
+   %arrayidx = OpInBoundsPtrAccessChain %_ptr_CrossWorkgroup_ushort %26 %27
+         %29 = OpLoad %ushort %arrayidx Aligned 2
+       %conv = OpSConvert %uint %29
+         %31 = OpLoad %_ptr_CrossWorkgroup_ushort %spec_constant_value_addr Aligned 4
+         %32 = OpLoad %ushort %31 Aligned 2
+      %conv1 = OpSConvert %uint %32
+        %add = OpIAdd %uint %conv %conv1
+      %conv2 = OpUConvert %ushort %add
```
This could be simplified by removing the OpSConvert and OpUConvert instructions.

StuartDBrady commented on test_conformance/spirv_new/spirv_asm/op_spec_constant_short_simple.spvasm64:
```diff
@@ -0,0 +1,73 @@
+         %27 = OpLoad %ulong %id Aligned 8
+   %arrayidx = OpInBoundsPtrAccessChain %_ptr_CrossWorkgroup_ushort %26 %27
+         %29 = OpLoad %ushort %arrayidx Aligned 2
+       %conv = OpSConvert %uint %29
+         %32 = OpLoad %_ptr_CrossWorkgroup_ushort %spec_constant_value_addr Aligned 8
+         %33 = OpLoad %ushort %32 Aligned 2
+      %conv1 = OpSConvert %uint %33
+        %add = OpIAdd %uint %conv %conv1
+      %conv2 = OpUConvert %ushort %add
```
This could be simplified by removing the OpSConvert and OpUConvert instructions.

StuartDBrady commented on test_conformance/spirv_new/spirv_asm/op_spec_constant_uchar_simple.spvasm64:
```diff
@@ -0,0 +1,71 @@
+         %26 = OpLoad %ulong %id Aligned 8
+   %arrayidx = OpInBoundsPtrAccessChain %_ptr_CrossWorkgroup_uchar %25 %26
+         %28 = OpLoad %uchar %arrayidx Aligned 1
+       %conv = OpUConvert %uint %28
+         %31 = OpLoad %_ptr_CrossWorkgroup_uchar %spec_constant_value_addr Aligned 8
+         %32 = OpLoad %uchar %31 Aligned 1
+      %conv1 = OpUConvert %uint %32
+        %add = OpIAdd %uint %conv %conv1
+      %conv2 = OpUConvert %uchar %add
```
This could be simplified by removing the OpUConvert instructions.

StuartDBrady commented on test_conformance/spirv_new/spirv_asm/op_spec_constant_uchar_simple.spvasm32:
```diff
@@ -0,0 +1,69 @@
+         %26 = OpLoad %uint %id Aligned 4
+   %arrayidx = OpInBoundsPtrAccessChain %_ptr_CrossWorkgroup_uchar %25 %26
+         %28 = OpLoad %uchar %arrayidx Aligned 1
+       %conv = OpUConvert %uint %28
+         %30 = OpLoad %_ptr_CrossWorkgroup_uchar %spec_constant_value_addr Aligned 4
+         %31 = OpLoad %uchar %30 Aligned 1
+      %conv1 = OpUConvert %uint %31
+        %add = OpIAdd %uint %conv %conv1
+      %conv2 = OpUConvert %uchar %add
```
This could be simplified by removing the OpUConvert instructions.

StuartDBrady commented on test_conformance/spirv_new/spirv_asm/op_spec_constant_ushort_simple.spvasm32:
```diff
@@ -0,0 +1,71 @@
+         %27 = OpLoad %uint %id Aligned 4
+   %arrayidx = OpInBoundsPtrAccessChain %_ptr_CrossWorkgroup_ushort %26 %27
+         %29 = OpLoad %ushort %arrayidx Aligned 2
+       %conv = OpUConvert %uint %29
+         %31 = OpLoad %_ptr_CrossWorkgroup_ushort %spec_constant_value_addr Aligned 4
+         %32 = OpLoad %ushort %31 Aligned 2
+      %conv1 = OpUConvert %uint %32
+        %add = OpIAdd %uint %conv %conv1
+      %conv2 = OpUConvert %ushort %add
```
This could be simplified by removing the OpUConvert instructions.

StuartDBrady commented on test_conformance/spirv_new/spirv_asm/op_spec_constant_ushort_simple.spvasm64:
```diff
@@ -0,0 +1,73 @@
+         %27 = OpLoad %ulong %id Aligned 8
+   %arrayidx = OpInBoundsPtrAccessChain %_ptr_CrossWorkgroup_ushort %26 %27
+         %29 = OpLoad %ushort %arrayidx Aligned 2
+       %conv = OpUConvert %uint %29
+         %32 = OpLoad %_ptr_CrossWorkgroup_ushort %spec_constant_value_addr Aligned 8
+         %33 = OpLoad %ushort %32 Aligned 2
+      %conv1 = OpUConvert %uint %33
+        %add = OpIAdd %uint %conv %conv1
+      %conv2 = OpUConvert %ushort %add
```
This could be simplified by removing the OpUConvert instructions.

gwawiork commented:
@StuartDBrady I've fixed recent review remarks. The changes in summary:
* no redundant OpUConvert OpSConvert
* no redundant OpName
* avoid using program that has already kernel program attached
* remove not used buffer
* Simplified test case - instead of buffers use scalar variable

StuartDBrady commented on test_conformance/spirv_new/test_op_spec_constant.cpp:
```diff
@@ -1,5 +1,5 @@
 /******************************************************************
-Copyright (c) 2016 The Khronos Group Inc. All Rights Reserved.
+Copyright (c) 2020 The Khronos Group Inc. All Rights Reserved.
```
Should we only be _adding_ new years? @kpet?

StuartDBrady commented on test_conformance/spirv_new/test_op_spec_constant.cpp:
As I previously explained, this does not work, as `get_program_with_ir()` calls `clBuildProgram()`.

StuartDBrady commented:
Looks okay to me, assuming @kpet is happy with the resolution of the changes he requested.

StuartDBrady commented:
@gwawiork, this needs a rebase before it can be merged.

gwawiork commented:
@StuartDBrady. Thanks a lot for review. After rebase I had to add additional fixing commit. Additionally I've added setting minimal version type as 2.2 for bool type tests

kpet commented:
This now needs rebasing again.

bashbaug commented on test_conformance/spirv_new/test_op_spec_constant.cpp:
Minor: notEqual doesn't seem to be used?

bashbaug commented on test_conformance/spirv_new/test_op_spec_constant.cpp:
Minor: consider checking the return value from `clFinish`.  Or, could it be omitted entirely?  The subsequent call to `clEnqueueReadBuffer` is blocking.

kpet reviewed (approved):
Agree with Ben here. This PR could use some more improvements but it's better to merge it as is to get forward progress and address the remaining comments later.

kpet commented:
Created #911 to cover the remaining comments, merging.
